src/main.c

```diff
@@ -0,0 +1,262 @@
+ *
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "board_api.h"
+#include "uf2.h"
+#include "tusb.h"
+
+//--------------------------------------------------------------------+
+// MACRO CONSTANT TYPEDEF PROTYPES
+//--------------------------------------------------------------------+
+//#define USE_DFU_BUTTON    1
+
+// Enter DFU magic
+#define DBL_TAP_MAGIC             0xf01669ef
+
+// Skip double tap delay detction
+#define DBL_TAP_MAGIC_QUICK_BOOT  0xf02669ef
+
+// timeout for double tap detection
+#define DBL_TAP_DELAY             500
+
+uint8_t const RGB_USB_UNMOUNTED[] = { 0xff, 0x00, 0x00 }; // Red
+uint8_t const RGB_USB_MOUNTED[]   = { 0x00, 0xff, 0x00 }; // Green
+uint8_t const RGB_WRITING[]       = { 0xcc, 0x66, 0x00 };
+uint8_t const RGB_DOUBLE_TAP[]    = { 0x80, 0x00, 0xff }; // Purple
+uint8_t const RGB_UNKNOWN[]       = { 0x00, 0x00, 0x88 }; // for debug
+uint8_t const RGB_OFF[]           = { 0x00, 0x00, 0x00 };
+
+//--------------------------------------------------------------------+
+//
+//--------------------------------------------------------------------+
+
+// defined by linker script
+extern uint32_t _board_dfu_dbl_tap[];
+static volatile uint32_t _timer_count = 0;
+
+// return true if start DFU mode, else App mode
+static bool check_dfu_mode(void)
+{
+  // Check if app is valid
+  if ( !board_app_valid() ) return true;
+
+#if USE_DFU_DOUBLE_TAP
+//  TU_LOG1_HEX(_board_dfu_dbl_tap);
+//  TU_LOG1_HEX(_board_dfu_dbl_tap[0]);
+
+  // App want to reboot quickly
+  if (_board_dfu_dbl_tap[0] == DBL_TAP_MAGIC_QUICK_BOOT)
+  {
+    _board_dfu_dbl_tap[0] = 0;
+    return false;
+  }
+
+  if (_board_dfu_dbl_tap[0] == DBL_TAP_MAGIC)
+  {
+    // Double tap occurred
+    _board_dfu_dbl_tap[0] = 0;
+    return true;
+  }
+
+  // Register our first reset for double reset detection
+  _board_dfu_dbl_tap[0] = DBL_TAP_MAGIC;
+  _timer_count = 0;
+
+  // Turn on LED/RGB for visual indicator
+  board_led_write(0xff);
+  board_rgb_write(RGB_DOUBLE_TAP);
+
+  // delay a fraction of second if Reset pin is tap during this delay --> we will enter dfu
+  board_timer_start(10);
+  while(_timer_count < DBL_TAP_DELAY/10) {}
+  board_timer_stop();
+
+  // Turn off indicator
+  board_led_write(0x00);
+
+  _board_dfu_dbl_tap[0] = 0;
+#endif
+
+  return false;
+}
+
+int main(void)
+{
+  board_init();
+
+  // if not DFU mode, jump to App
+  if ( !check_dfu_mode() )
+  {
+    board_app_jump();
+    while(1) {}
+  }
+
+  board_dfu_init();
+  board_flash_init();
+  uf2_init();
+
+  tusb_init();
+
+  indicator_set(STATE_BOOTLOADER_STARTED);
+
+#if USE_SCREEN
+  screen_init();
+  screen_draw_drag();
+#endif
+
+#if CFG_TUSB_OS == OPT_OS_NONE
+  while(1)
+  {
+    tud_task();
+  }
+#endif
+}
+
+//--------------------------------------------------------------------+
+// Device callbacks
+//--------------------------------------------------------------------+
+
+// Invoked when device is mounted

```

Comment:
Could we say "plugged in" and "unplugged"? "mounted" is confusing because it doesn't relate to the mass storage drive.

Reply:
done